File: benchmarks/realworld/edgebmp/C/edgebmp.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <math.h>
#include <time.h>

#pragma pack(push, 1)
typedef struct {
    uint16_t bfType;
    uint32_t bfSize;
    uint16_t bfReserved1, bfReserved2;
    uint32_t bfOffBits;
} BITMAPFILEHEADER;

typedef struct {
    uint32_t biSize;
    int32_t  biWidth, biHeight;
    uint16_t biPlanes, biBitCount;
    uint32_t biCompression, biSizeImage;
    int32_t  biXPelsPerMeter, biYPelsPerMeter;
    uint32_t biClrUsed, biClrImportant;
} BITMAPINFOHEADER;
#pragma pack(pop)

double now_sec() {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec + ts.tv_nsec / 1e9;
}

int clamp(int val, int min, int max) {
    if (val < min) return min;
    if (val > max) return max;
    return val;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        printf("Usage: %s <image.bmp>\n", argv[0]);
        return 1;
    }

    FILE *fp = fopen(argv[1], "rb");
    if (!fp) {
        perror("fopen");
        return 1;
    }

    BITMAPFILEHEADER fileHeader;
    BITMAPINFOHEADER infoHeader;

    fread(&fileHeader, sizeof(fileHeader), 1, fp);
    fread(&infoHeader, sizeof(infoHeader), 1, fp);

    if (fileHeader.bfType != 0x4D42 || infoHeader.biBitCount != 24) {
        printf("Only 24-bit BMP supported.\n");
        fclose(fp);
        return 1;
    }

    int width = infoHeader.biWidth;
    int height = abs(infoHeader.biHeight);
    int padding = (4 - (width * 3) % 4) % 4;

    uint8_t *gray = malloc(width * height);
    fseek(fp, fileHeader.bfOffBits, SEEK_SET);

    for (int y = height - 1; y >= 0; y--) {
        for (int x = 0; x < width; x++) {
            uint8_t bgr[3];
            fread(bgr, 1, 3, fp);
            uint8_t grayVal = (uint8_t)(0.299 * bgr[2] + 0.587 * bgr[1] + 0.114 * bgr[0]);
            gray[y * width + x] = grayVal;
        }
        fseek(fp, padding, SEEK_CUR);
    }
    fclose(fp);

    // Sobel Kernel
    int gx[3][3] = {
        {-1, 0, 1}, {-2, 0, 2}, {-1, 0, 1}
    };
    int gy[3][3] = {
        {1, 2, 1}, {0, 0, 0}, {-1, -2, -1}
    };

    double start = now_sec();

    uint8_t *output = malloc(width * height);
    for (int y = 1; y < height - 1; y++) {
        for (int x = 1; x < width - 1; x++) {
            int sumX = 0, sumY = 0;
            for (int dy = -1; dy <= 1; dy++) {
                for (int dx = -1; dx <= 1; dx++) {
                    uint8_t val = gray[(y + dy) * width + (x + dx)];
                    sumX += val * gx[dy + 1][dx + 1];
                    sumY += val * gy[dy + 1][dx + 1];
                }
            }
            int mag = sqrt(sumX * sumX + sumY * sumY);
            output[y * width + x] = clamp(mag, 0, 255);
        }
    }

    double end = now_sec();

    printf("Edge detection time: %.6f seconds\n", end - start);

    FILE *out = fopen("output.raw", "wb");
// 写入 BMP 文件头（自定义 24bit 灰度 BMP）

    fwrite(output, 1, width * height, out);
    fclose(out);

    free(gray);
    free(output);
    return 0;
}

File: benchmarks/realworld/edgebmp/Python/edgebmp.py
import struct
import math
import time

def now_sec():
    return time.time()

def clamp(val, min_val, max_val):
    return min(max(val, min_val), max_val)

def main():
    import sys
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <image.bmp>")
        return 1

    with open(sys.argv[1], 'rb') as fp:
        # Read BITMAPFILEHEADER
        data = fp.read(14)
        bfType, bfSize, bfReserved1, bfReserved2, bfOffBits = \
            struct.unpack('<HIHHI', data)
        
        if bfType != 0x4D42:
            print("Only BMP supported.")
            return 1

        # Read BITMAPINFOHEADER
        data = fp.read(40)
        (biSize, biWidth, biHeight, biPlanes, biBitCount, biCompression,
         biSizeImage, biXPelsPerMeter, biYPelsPerMeter, biClrUsed,
         biClrImportant) = struct.unpack('<IiiHHIIiiII', data)
        
        if biBitCount != 24:
            print("Only 24-bit BMP supported.")
            return 1

        width = biWidth
        height = abs(biHeight)
        padding = (4 - (width * 3) % 4) % 4

        gray = bytearray(width * height)
        fp.seek(bfOffBits)

        for y in range(height - 1, -1, -1):
            for x in range(width):
                bgr = fp.read(3)
                gray_val = int(0.299 * bgr[2] + 0.587 * bgr[1] + 0.114 * bgr[0])
                gray[y * width + x] = gray_val
            fp.seek(padding, 1)

    # Sobel Kernel
    gx = [[-1, 0, 1], [-2, 0, 2], [-1, 0, 1]]
    gy = [[1, 2, 1], [0, 0, 0], [-1, -2, -1]]

    start = now_sec()

    output = bytearray(width * height)
    for y in range(1, height - 1):
        for x in range(1, width - 1):
            sum_x = sum_y = 0
            for dy in range(-1, 2):
                for dx in range(-1, 2):
                    val = gray[(y + dy) * width + (x + dx)]
                    sum_x += val * gx[dy + 1][dx + 1]
                    sum_y += val * gy[dy + 1][dx + 1]
            mag = int(math.sqrt(sum_x**2 + sum_y**2))
            output[y * width + x] = clamp(mag, 0, 255)

    end = now_sec()
    print(f"Edge detection time: {end - start:.6f} seconds")

    # Optional: write output.raw
    with open('output.raw', 'wb') as f:
        f.write(output)

if __name__ == "__main__":
    main()
